src/storage/archive_index.rs

```diff
@@ -0,0 +1,106 @@
+use crate::storage::{compression::CompressionAlgorithm, FileRange};
+use failure::Error;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io;
+use std::path::{Path, PathBuf};
+
+#[derive(Deserialize, Serialize)]
+pub struct FileInfo {
+    range: FileRange,
+    compression: CompressionAlgorithm,
+}
+
+impl FileInfo {
+    pub fn range(&self) -> FileRange {
+        self.range.clone()
+    }
+    pub fn compression(&self) -> CompressionAlgorithm {
+        self.compression
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct Index {
+    files: HashMap<PathBuf, FileInfo>,
+}
+
+impl Index {
+    pub fn load(reader: impl io::Read) -> Result<Index, Error> {
+        serde_json::from_reader(reader).map_err(|_| failure::err_msg("deserialization error"))
```

Comment:
that was also on my list to figure out. 

Both `.context` and directly returning the result don't compile because (I assume) something in the trait bounds on the return type of `from_reader`.

Reply:
( I wanted to make it work first)